spartan_parallel/src/sparse_mlpoly.rs

```diff
@@ -564,62 +494,60 @@ impl MultiSparseMatPolynomialAsDense {
 }
 
 #[derive(Debug)]
-struct ProductLayer {
-  init: ProductCircuit,
-  read_vec: Vec<ProductCircuit>,
-  write_vec: Vec<ProductCircuit>,
-  audit: ProductCircuit,
+struct ProductLayer<S: SpartanExtensionField> {
+  init: ProductCircuit<S>,
+  read_vec: Vec<ProductCircuit<S>>,
+  write_vec: Vec<ProductCircuit<S>>,
+  audit: ProductCircuit<S>,
 }
 
 #[derive(Debug)]
-struct Layers {
-  prod_layer: ProductLayer,
+struct Layers<S: SpartanExtensionField> {
+  prod_layer: ProductLayer<S>,
 }
 
-impl Layers {
+impl<S: SpartanExtensionField> Layers<S> {
   fn build_hash_layer(
-    eval_table: &[Scalar],
-    addrs_vec: &[DensePolynomial],
-    derefs_vec: &[DensePolynomial],
-    read_ts_vec: &[DensePolynomial],
-    audit_ts: &DensePolynomial,
-    r_mem_check: &(Scalar, Scalar),
+    eval_table: &[S],
+    addrs_vec: &[DensePolynomial<S>],
+    derefs_vec: &[DensePolynomial<S>],
+    read_ts_vec: &[DensePolynomial<S>],
+    audit_ts: &DensePolynomial<S>,
+    r_mem_check: &(S, S),
   ) -> (
-    DensePolynomial,
-    Vec<DensePolynomial>,
-    Vec<DensePolynomial>,
-    DensePolynomial,
+    DensePolynomial<S>,
+    Vec<DensePolynomial<S>>,
+    Vec<DensePolynomial<S>>,
+    DensePolynomial<S>,
   ) {
     let (r_hash, r_multiset_check) = r_mem_check;
 
     //hash(addr, val, ts) = ts * r_hash_sqr + val * r_hash + addr
-    let r_hash_sqr = r_hash * r_hash;
-    let hash_func = |addr: &Scalar, val: &Scalar, ts: &Scalar| -> Scalar {
-      ts * r_hash_sqr + val * r_hash + addr
-    };
+    let r_hash_sqr = *r_hash * *r_hash;
+    let hash_func = |addr: &S, val: &S, ts: &S| -> S { *ts * r_hash_sqr + *val * *r_hash + *addr };
```

Comment:
This closure can be boosted by by changing the type of `addr, ts` from `S` to `S::BaseField` as extension field element multiply by base field element `e*b` (https://github.com/scroll-tech/ceno-Goldilocks/blob/master/src/extfield.rs#L24) is much faster than `e * e`